x/emissions/keeper/keeper.go

```diff
@@ -1004,64 +1029,79 @@ func (k *Keeper) AppendInference(
+		if err != nil {
+			return errorsmod.Wrap(err, "error getting lowest score from all inferers")
+		}
+		err = k.SetLowestInfererScoreEma(ctx, topic.Id, lowestEmaScore)
+		if err != nil {
+			return errorsmod.Wrap(err, "error setting lowest inferer score ema")
+		}
 	}
-	// Update EMA score for the lowest score inferer
-	// This conditional decides who that lowest score inferer is
-	if previousEmaScore.Score.Gt(lowScore.Score) {
+
+	if previousEmaScore.Score.Gt(lowestEmaScore.Score) {
 		// Update EMA score for the lowest score inferer, who is not the current inferer
 		err = k.CalcAndSaveInfererScoreEmaWithLastSavedTopicQuantile(
 			ctx,
 			topic,
 			nonceBlockHeight,
-			inferences.Inferences[lowScoreIndex].Inferer,
+			lowestEmaScore,
 		)
 		if err != nil {
 			return errorsmod.Wrap(err, "error calculating and saving inferer score ema with last saved topic quantile")
 		}
-		// Remove the lowest score inferer from the list
-		inferences.Inferences = append(inferences.Inferences[:lowScoreIndex], inferences.Inferences[lowScoreIndex+1:]...)
-		// Append the current inferer to the list
-		inferences.Inferences = append(inferences.Inferences, inference)
-		return k.InsertInferences(ctx, topic.Id, nonceBlockHeight, inferences)
+		// Remove inferer with lowest score
+		err = k.RemoveActiveInferer(ctx, topic.Id, lowestEmaScore.Address)
+		if err != nil {
+			return errorsmod.Wrap(err, "error removing active inferer")
+		}
+		// Add new active inferer
+		err = k.AddActiveInferer(ctx, topic.Id, inference.Inferer)
+		if err != nil {
+			return errorsmod.Wrap(err, "error adding active inferer")
+		}
+		// Calculate new lowest score with updated infererAddresses
+		err = UpdateLowestScoreFromInfererAddresses(ctx, k, topic.Id, workerAddresses, inference.Inferer, lowestEmaScore.Address)
+		if err != nil {
+			return errorsmod.Wrap(err, "error getting low score from all inferences")
+		}
+		return k.InsertInference(ctx, topic.Id, *inference)
 	} else {
 		// Update EMA score for the current inferer, who is the lowest score inferer
-		err = k.CalcAndSaveInfererScoreEmaWithLastSavedTopicQuantile(ctx, topic, nonceBlockHeight, inference.Inferer)
+		err = k.CalcAndSaveInfererScoreEmaWithLastSavedTopicQuantile(ctx, topic, nonceBlockHeight, previousEmaScore)
 		if err != nil {
 			return errorsmod.Wrap(err, "error calculating and saving inferer score ema with last saved topic quantile")
 		}
 	}
 	return nil
 }
 
-// Insert a complete set of inferences for a topic/block. Overwrites previous ones.
-func (k *Keeper) InsertInferences(
+// Insert an inference for a specific topic
+func (k *Keeper) InsertInference(
+	ctx context.Context,
+	topicId TopicId,
+	inference types.Inference,
+) error {
+	err := inference.Validate()
+	if err != nil {
+		return errorsmod.Wrap(err, "inference in list is invalid")
+	}
+	key := collections.Join(topicId, inference.Inferer)
+	return k.inferences.Set(ctx, key, inference)
+}
```

Comment:
I can't find where the `inferences` store is cleared. 
Seems like it should have all of its entries deleted before getting a new epoch in, right? 
(same for `forecasts`)

Reply:
added!